File: 9012.py
import sys

n = int(sys.stdin.readline())

for _ in range(n):
    temp = sys.stdin.readline().strip()
    stack = []
    isBreak = False
    for s in temp:
        if s == "(":
            stack.append(s)
        else:
            if stack:
                if stack.pop() != "(":
                    print("NO")
                    isBreak = True
                    break
            else:
                print("NO")
                isBreak = True
                break
    if not isBreak:
        if stack:
            print("NO")
        else:
            print("YES") 

File: 9012.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int main(){
    int n,temp;
    string a;
    cin >> n;

    for (int i=0; i<n; i++){
        cin >> a;
        temp = 0;
        for(auto c:a){
            if(c == '('){
                temp += 1;
            }
            else{
                temp -= 1;
            }
            if(temp < 0) break;
        } 
        if (temp == 0){
            cout << "YES\n";
        }
        else{
            cout << "NO\n";
        }
    }
}